kvserver,cli,roachtest,sql: introduce a fully decommissioned bit

irfansharif opened the pull request:
Ignore first commit, cherry-picked from @andreimatei's #49135.

---

This commit introduces a fully decommissioned bit to Cockroach.
Previously our Liveness schema only contained a decommissioning bool,
with consequently no ability to disamiguate between a node currently
undergoing decommissioning, and a node that was fully decommissioned. We
used some combination of store dead threshold to surface, in our UI,
"fully decommissioned" nodes, but it was never quite so. We need this
specificity for the Connect RPC.

We wire up the new CommissionStatus enum that's now part of the liveness
record. In doing so it elides usage of the decommissioning bool used in
v20.1. We're careful to maintain an on-the-wire representation of the
Liveness record that will be understood by v20.1 nodes (see
Liveness.EnsureCompatible for details).

We repurpose the AdminServer.Decommission RPC (which is now a misnomer,
should be thought of being named Commission instead) to persist
CommissionStatus states to KV through the lifetime of a node
decommissioning/recommissioning. See cli/node.go for where that's done.
For recommissioning a node, it suffices to simply persist a COMMISSIONED
state. When decommissioning a node, since it's a longer running process,
we first persist an in-progress DECOMMISSIONING state, and once we've
moved off all the Replicas in the node, we finalize the decommissioning
process by persisting the DECOMMISSIONED state.

When transitioning between CommissionStatus states, we CPut against
what's already there, disallowing illegal state transitions. The
appropriate error codes are surfaced back to the user. An example would
be in attempting to recommission a fully decommissioned node, in which
case we'd error out with the following:

> command failed: illegal commission status transition: can only
> recommission a decommissioning node; n4 found to be decommissioned

Note that this is a behavioral change for `cockroach node recommission`.
Previously it was able to recommission any "decommissioned" node,
regardless of how long ago it's was removed from the cluster. Now
recommission serves to only cancel an accidental decommissioning process
that wasn't finalized.

The 'decommissioning' column in crdb_internal.gossip_liveness is now
powered by this new CommissionStatus, and we introduce a new
commission_status column to it. We also introduce the same column to
the output generated by `cockroach node status --decommission`. The
is_decommissioning column still exists, but is also powered by the
CommissionStatus now.

We also iron out the events plumbed into system.eventlog: it now has a
dedicated event for "node decommissioning".

Release note (general change): `cockroach node recommission` has new
semantics.Previously it was able to recommission any decommissioned node,
regardless of how long ago it's was fully removed from the cluster. Now
recommission serves to only cancel an accidental decommissioning process
that wasn't finalized.

Release note (cli change): We introduce a commission_status column to
the output generated by `cockroach node status --decommission`. It
should be used in favor of the is_decommissioning column going forward.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50329)
<!-- Reviewable:end -->

tbg reviewed (commented):
Thanks Irfan! Just got through the first round of review. Summarizing the main work items that I see (they're also in individual comments):

- it seems better to reuse the `decommissioning` bool (make it an enum). I saw all of this compatibility logic and it looks way more complicated than I hope it needs to be. Reusing the field, all you have to do is put a cluster version check that we're never writing enum values other than `0` and `1` until we have the new version active, and none of the consumers of liveness statuses really need special logic.
- worried that `commission` is not an intuitive choice. My strawman is to name this the "membership status" which is either "active", "decommissioning", "decommissioned". I think that users seeing that in their output would intuitively know what that is more than "commission status: commissioning".

Also, in the diagram, it looks like there's an arrow from "decommissioned" to "decommissioning" which should not be there.

It may also be a good time to write a mixed-version decommissioning test, perhaps as a replacement to the acceptance test which dropped out due to needing (why?) more nodes. This could be kept relatively simple, set up a four node cluster in which two nodes are new and two are old, then pick a random node to decommission and check that it goes through without a hitch (using the old binary as cli, I think you'll have a hard time getting the new one to work since it expects the new column to exist). `runVersionUpgrade` is a good template to follow for mixed-version stuff. (I'd be fine with not doing all of the cli output parsing which is super nasty but I don't see an obvious better way).

You may want to pull out the deprecation of the nodelivenessstatus proto into a separate PR so that it's out of your hair here. I would prefer to keep this minimal (just a rename) and to cut down on its uses only when you are ready to focus on it (i.e. probably not now).

> Reviewed 11 of 11 files at r1, 27 of 27 files at r2, 2 of 2 files at r3, 19 of 19 files at r4, 4 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-:-MA5n2vc9pvd-wGav6-B:b-34emby)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @knz)

___
*[pkg/cli/demo_cluster.go, line 375 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5nA915hpzgTPRijuU:-MA5nA915hpzgTPRijuV:bttvoj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/demo_cluster.go#L375)):*
> ```Go
> 	req := &serverpb.CommissionRequest{
> 		NodeIDs:          []roachpb.NodeID{nodeID},
> 		CommissionStatus: kvserverpb.CommissionStatusFromBooleanForm(decommissioning),
> ```

why `Form`? Isn't `CommissionStatusFromBool` good enough?
___
*[pkg/cli/node.go, line 66 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5nG9aAl2XQtT5_C0L:-MA5nG9aAl2XQtT5_C0M:b-eydts3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L66)):*
> ```Go
> 		conn,
> 		makeQuery(`SELECT node_id FROM crdb_internal.gossip_liveness
>                WHERE commission_status = 'commissioned' OR split_part(expiration,',',1)::decimal > now()::decimal`),
> ```

Add a release note that the 20.2 cli `node` family of commands will not work with older versions, but that older versions will work against 20.2.
___
*[pkg/cli/node.go, line 340 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5oCyfBjNiAR6UDeJE:-MA5oCyfBjNiAR6UDeJF:b-sek8xg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L340)):*
> ```Go
> 		req := &serverpb.CommissionRequest{
> 			NodeIDs:          nodeIDs,
> 			CommissionStatus: kvserverpb.CommissionStatus_DECOMMISSIONING,
> ```

If a 20.2 cli were used against a 20.1 node, it would always recommission, which is bad.
I think this problem goes away if you re-use the field as I suggested in another comment.
___
*[pkg/cli/node.go, line 361 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5oxkX88YCveznVAeI:-MA5oxkY2d3MmZJF5FvK:bbrp6s4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L361)):*
> ```Go
> 		for _, status := range resp.Status {
> 			replicaCount += status.ReplicaCount
> 			allDecommissioning = allDecommissioning && status.CommissionStatus.Decommissioning() ||
> ```

Uh, oh. `false && false || true == true`, so this is a bug. You want `false && (false || true)`
___
*[pkg/cli/node.go, line 364 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5pOCQ4t5D7hGdzp7x:-MA5pOCQ4t5D7hGdzp7y:bnb2war) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L364)):*
> ```Go
> 				status.CommissionStatus.Decommissioned()
> 		}
> 		if replicaCount == 0 {
> ```

You would never want to enter this if `!allDecommissioning`, no?
___
*[pkg/cli/node.go, line 365 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5ojCa4GgbWgjM1snT:-MA5oPR7-23rBz9Dwa0E:b-iwuqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L365)):*
> ```Go
> 		}
> 		if replicaCount == 0 {
> 			// We now mark the node as fully decommissioned.
> ```

node(s)
___
*[pkg/cli/node.go, line 389 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5pGvODpM2GgrIgy5n:-MA5pGvODpM2GgrIgy5o:b-z7i93u) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L389)):*
> ```Go
> 		}
> 		if wait == nodeDecommissionWaitNone {
> 			// We don't bother with marking the node as fully decommissioned.
> ```

But you may already have done it? I think you need to make sure you never do it in that case.

Let's give this method a real close look, since it seems really easy to introduce bugs here.
___
*[pkg/cli/node.go, line 419 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5podeEyakZx_SE4hw:-MA5podf2A9vo1adpPO0:ba9hakg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L419)):*
> ```Go
> 			strconv.FormatBool(node.IsLive),
> 			strconv.FormatInt(node.ReplicaCount, 10),
> 			strconv.FormatBool(node.CommissionStatus.Decommissioning() || node.CommissionStatus.Decommissioned()),
> ```

All of these too will be better off with the enum reusing the bool field. Right now if you use a new cli against old node you'll just get wrong answers. I'm ok breaking compatibility here but we shouldn't silently return wrong data. Problem will disappear with enum reuse, luckily.
___
*[pkg/cli/node.go, line 446 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5qK380yR6ENzPK5Kl:-MA5qK380yR6ENzPK5Km:b-ba2qo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L446)):*
> ```Go
> // messages generated by cockroach.
> func extractCommissionErr(errMsg string) string {
> 	idx := strings.Index(errMsg, kvserverpb.ErrIllegalCommissionStatusTransition.Error())
> ```

This returns -1 if not found, in which case you'll explode below.

I see now that you guard against that at the caller, but move that check in here. If you can't match, return "".
___
*[pkg/cli/node.go, line 474 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5qd3v34cfDlOpUXsj:-MA5qd3v34cfDlOpUXsk:b-k939mj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L474)):*
> ```Go
> 		// If it's a specific illegal commission transition error, we try to
> 		// surface a more readable message to the user.
> 		if strings.Contains(err.Error(), kvserverpb.ErrIllegalCommissionStatusTransition.Error()) {
> ```

Isn't there a better way to solve this? You can return a structured grpc error:

```go
err := status.Error(codes.FailedPrecondition, "something something blah blah")
```

and then later

```
if s, ok := status.FromError(err); ok && s.Code() == codes.FailedPrecondition {
	fmt.Println(s.Message())
}
```
___
*[pkg/cli/interactive_tests/test_demo_node_cmds.tcl, line 70 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5sCUf7JqtHv8gMpON:-MA5sCUf7JqtHv8gMpOO:b-4f8fbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/interactive_tests/test_demo_node_cmds.tcl#L70)):*
> ```Tcl
> 
> send "select node_id, draining, decommissioning, commission_status from crdb_internal.gossip_liveness ORDER BY node_id;\r"
> eexpect "1 |  false   |      false      | commissioned"
> ```

It's been lingering in my mind throughout the review, and I think for a good reason - I'm not sure `commissioned` really is the most intuitive choice here. It looks like a verb - like we commissioned a node, i.e. we ordered one, it's bound to arrive any time soon. I feel that this might be confusing to users especially when we have some very obvious choices available, like `active`. Consider switching to that.
___
*[pkg/cli/interactive_tests/test_multiple_nodes.tcl, line 24 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5mS6H0d-BqHDeG8cN:-MA5mS6H0d-BqHDeG8cO:b-vq5sn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/af0f7bb31be91b0fe61a30d82054223293c6c597/pkg/cli/interactive_tests/test_multiple_nodes.tcl#L24)):*
> ```Tcl
> 
> 
> start_test "Check that a double decommission prints out a warning"
> ```

Do we want to maintain these warnings though? In that case it seems wise to remove the code only when we're ready to replace it with new code. It seems that this PR is trying to do two things at once - add the decommission bit and deprecating the liveness status thing, but it's not really following through on the latter, so I would prefer if we kept that to the minimum and followed through in a separate PR.
___
*[pkg/cmd/roachtest/acceptance.go, line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5sh4r1Bx1hAigt-Bx:-MA5sh4r1Bx1hAigt-By:b3kre) ([raw file](https://github.com/cockroachdb/cockroach/blob/af0f7bb31be91b0fe61a30d82054223293c6c597/pkg/cmd/roachtest/acceptance.go#L38)):*
> ```Go
> 		{name: "build-analyze", fn: runBuildAnalyze},
> 		{name: "cli/node-status", fn: runCLINodeStatus},
> 		{name: "decommission", fn: runDecommissionAcceptance},
> ```

?
___
*[pkg/cmd/roachtest/decommission.go, line 253 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5t_wpDRVEUBJq6tqV:-MA5t_wpDRVEUBJq6tqW:b-4ch8wj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cmd/roachtest/decommission.go#L253)):*
> ```Go
> 		r.Add(testSpec{
> 			// This test was originally an acceptance test, we preserve its name.
> 			Name:    "acceptance/decommission",
> ```

There's no issue open for it, so let's rename it now.
___
*[pkg/cmd/roachtest/decommission.go, line 279 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5tk0Z10G-FgJ0_lnt:-MA5tk0Z10G-FgJ0_lnu:blt7dxf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cmd/roachtest/decommission.go#L279)):*
> ```Go
> // runDecommissionAcceptance tests a bunch of node
> // decommissioning/recommissioning procedures, all the while checking for
> // replica movement and appropriate commission status detection behavior.
> ```

It's great that you're updating this test, but I got a little lost in what it's actually doing now (and why it needs so many nodes). Could you provide a high-level overview of what operations it carries out here in the top comment? I know this may go stale at some point but it will still be better than having to trawl through the code every time.
In particular, are we at any point decommissioning multiple nodes at once (i.e. with one cli invocation)? If not that would be good to add now that we can.

Did you verify that this still works against <= 20.1? I'd be sort of surprised if it did, but we only have one decommissioning test and so it will have to.
___
*[pkg/kv/kvserver/node_liveness.go, line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5uVFC5MYbOemAlR9Q:-MA5uVFC5MYbOemAlR9R:bi22wur) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/kv/kvserver/node_liveness.go#L49)):*
> ```Go
> 	// conditionally write the target commissioning status. It's safe to retry
> 	// when encountering this error.
> 	errChangeCommissionStatusFailed = errors.New("failed to change the commissioning status")
> ```

Similar concern about `commissioning status` (which pops up in a number of column headers). What about `membership` status? After all, decommissioning is all about removing members from a cluster and in the context of long-running operations, we are trying to reason about cluster membership in an even stricter sense (who *can* access the cluster as opposed to who is tracked as a member)
___
*[pkg/kv/kvserver/node_liveness.go, line 360 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5vN5wDZX07su7gHiZ:-MA5vN5wDZX07su7gHi_:brw67l) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/kv/kvserver/node_liveness.go#L360)):*
> ```Go
> 		if errors.Is(err, errChangeCommissionStatusFailed) {
> 			// Expected when epoch incremented, or when running in mixed version
> 			// clusters. It's safe to retry.
> ```

This shouldn't fail in mixed-version clusters. I don't think it would when you take my suggestion to merge the fields. The only caveat is that we are not allowed to mark as `decommissioned` but the `cli` should just special-case the mixed version error there and print a warning to the client (but succeed anyway).
___
*[pkg/kv/kvserver/node_liveness.go, line 382 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5vZlJ7z0frDYGF8rF:-MA5vZlJ7z0frDYGF8rG:b-84ydmm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/kv/kvserver/node_liveness.go#L382)):*
> ```Go
> 	}()
> 
> 	// Lets compute what our new liveness record should be.
> ```

Let's
___
*[pkg/kv/kvserver/node_liveness.go, line 395 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5viwi6YJg2NEQ9H3x:-MA5viwi6YJg2NEQ9H3y:bfi6far) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/kv/kvserver/node_liveness.go#L395)):*
> ```Go
> 		// We may have read a liveness record written by a v20.1 node, so we
> 		// reconcile.
> 		newLiveness.EnsureCompatible()
> ```

Just a heads up that I'm going to skip lightly over this mixed version stuff for now since I expect it to change.
___
*[pkg/kv/kvserver/node_liveness.go, line 438 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5vpBGD1l96sAzynBd:-MA5vpBGD1l96sAzynBe:b-odke43) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/kv/kvserver/node_liveness.go#L438)):*
> ```Go
> // liveness record. It has both the new and the old version of the proto.
> type livenessUpdate struct {
> 	new kvserverpb.Liveness
> ```

`new` is also a keyword, let's avoid that.
___
*[pkg/kv/kvserver/kvserverpb/liveness.go, line 71 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5wEBODVbUDQ9EgcRK:-MA5wEBODVbUDQ9EgcRL:b-kfxwny) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/kv/kvserver/kvserverpb/liveness.go#L71)):*
> ```Go
> // authoritative if not.
> //
> // TODO(irfansharif): Remove this once v20.2 is cut.
> ```

Could this really be removed? Nodes that never join the cluster with 20.2 will never update their liveness, so you'd have to retain some understanding of the old format ~forever (or write an explicit migration).
I think this problem will go away when you merge the fields.
___
*[pkg/kv/kvserver/kvserverpb/liveness.proto, line 51 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5kZYK0J_q096PRfsd:-MA5kZYL6ZQVcsONlFBH:b-4nzhkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cb13324e4f721178bbe942958e3c3d5dfa05b96/pkg/kv/kvserver/kvserverpb/liveness.proto#L51)):*
> ```protobuf
>   // fully decommissioned. It is deprecated in favor of CommissionStatus, and is
>   // slated for removal in v21.1.
>   bool deprecated_decommissioning = 5;
> ```

The protobuf encoding uses varint encoding for both bools and enum, so you don't need to add a new field. You can change the type and make sure that the enum has `0` as `Commissioned` and `1` as `decommissioning`. And of course you cannot use any other value unless the cluster version is active.
___
*[pkg/kv/kvserver/kvserverpb/liveness.proto, line 128 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5l8x2B5PAWGcoYEVT:-MA5l8x2B5PAWGcoYEVU:b-f4fzb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cb13324e4f721178bbe942958e3c3d5dfa05b96/pkg/kv/kvserver/kvserverpb/liveness.proto#L128)):*
> ```protobuf
> // across these states (LivenessStatus for e.g.) seems wholly arbitrary.
> enum NodeLivenessStatus {
>   DEPRECATED_UNKNOWN = 0;
> ```

Next time, make sure you put pure renames like that in a separate commit. There were three things going in on in this one which was mildly unfortunate.
___
*[pkg/server/server.go, line 1807 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5weOUBySENylSRirM:-MA5weOV2KojZ9269oO2:b7rp1bd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/server/server.go#L1807)):*
> ```Go
> 		eventType = sql.EventLogNodeDecommissioned
> 	} else if targetStatus.Commissioned() {
> 		eventType = sql.EventLogNodeCommissioned
> ```

And here you'd check the cluster version.
___
*[pkg/server/server.go, line 1809 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5wgc94GArAcHcx-k-:-MA5wgc94GArAcHcx-k0:b-cpsflw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/server/server.go#L1809)):*
> ```Go
> 		eventType = sql.EventLogNodeCommissioned
> 	} else {
> 		panic("unexpected target commission status")
> ```

We need to avoid panicking whenever we can. Here it's easy to return an error.
___
*[pkg/server/serverpb/admin.proto, line 879 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5lTKcDdW1SN1jRbgV:-MA5lTKcDdW1SN1jRbgW:bjgjms1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cb13324e4f721178bbe942958e3c3d5dfa05b96/pkg/server/serverpb/admin.proto#L879)):*
> ```protobuf
>   //
>   // The name of this method, Decommission, is a misnomer for legacy
>   // reasons. It should be thought of as `Commission` instead.
> ```

I would prefer if we just stuck with the old name throughout. `CommissionRequest` sounds like a request to "commission" a node, i.e. to mark it as commissioned, and thus doesn't really seem preferable to `DecommissionRequest` (which at least will be used for its name most of the time).
___
*[pkg/sql/crdb_internal.go, line 2742 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5wpnL7a23YKOocdOb:-MA5wpnL7a23YKOocdOc:bgeqzi6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/sql/crdb_internal.go#L2742)):*
> ```Go
>   draining            BOOL NOT NULL,
>   decommissioning     BOOL NOT NULL,
>   commission_status   STRING NOT NULL,
> ```

Looking at this, I think `membership: active` would have a good intuitive ring to it.
___
*[pkg/sql/event_log.go, line 90 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5wzac8bircHsWKSzQ:-MA5wzac8bircHsWKSzR:b-rkkh8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/sql/event_log.go#L90)):*
> ```Go
> 	// EventLogNodeDecommissioned is recorded when a node is marked as
> 	// decommissioned.
> 	EventLogNodeDecommissioned EventLogType = "node_decommissioned"
> ```

Is there a migration concern with this change? Not really right? They're only ever used as strings.


<!-- Sent from Reviewable.io -->

irfansharif commented:
Just an update here: I've addressed everything brought up so far (pushing the raw form to run CI) save for the mixed version test. I tried it out manually and seems to work cleanly now that we've upgraded the bool instead of introducing a new enum. I'm going to do some leg work tomorrow to clean it back up to be reviewable again (+ add the mixed version test), in addition to rebasing it atop https://github.com/cockroachdb/cockroach/pull/50478. I'll ping back on both issues once done.

irfansharif committed:
cli: improve help prompt for --wait={all,none} for node decommissioning

Now that we have a fully decommissioned bit, we clarify the mechanics of
how that is interfaced with through the `--wait` flag.

Release note (cli change): We slightly change the mechanics of how the
`--wait` flag, as used by `cockroach node decommission`, behaves.
Copying over from the help prompt:

```
  Specifies when to return during the decommissioning process. Takes any
  of the following values:

    - all   waits until all target nodes' replica counts have dropped to zero and
            marks the nodes as fully decommissioned. This is the default.
    - none  marks the targets as decommissioning, but does not wait for the
            replica counts to drop to zero before returning. If the replica counts
            are found to be zero, nodes are marked as fully decommissioned. Use
            when polling manually from an external system.
```

irfansharif committed:
kvserver,cli,roachtest,sql: introduce a decommissioned bit

This PR introduces a fully decommissioned bit to CRDB. Previously our
Liveness schema only contained a `decommissioning` bool, with
consequently no ability to disamiguate between a node currently
undergoing decommissioning, and a node that was fully decommissioned. We
used some combination of store dead threshold to surface, in our UI,
"fully decommissioned" nodes, but it was never quite so. We need this
specificity for the Connect RPC.

---

We wire up a new `MembershipStatus` enum that's now part of the liveness
record. In doing so it elides usage of the `decommissioning` bool used
in v20.1. We're careful to maintain an on-the-wire representation of the
Liveness record that will be understood by v20.1 nodes, and do so by
ensuring the encoding of the enum type is parsed into the semantically
equivalent v20.1 representation. Usage of the fully decommissioned bit
is gated behind a version flag. A future commit will introduce a
mixed-version roachtest testing cross-version compatibility. A future
commit will also re-register/unskip an improved version of the
`acceptance/decommission` roachtest.

We repurpose the `AdminServer.Decommission` RPC to persist
`MembershipStatus`es to KV through the lifetime of a node
decommissioning/recommissioning. See `cli/node.go` for where that's done.
For recommissioning a node, it suffices to simply persist an `ACTIVE`
status. When decommissioning a node, since it's a longer running process,
we first persist an in-progress `DECOMMISSIONING` status, and once we've
moved off all the Replicas in the node, we finalize the decommissioning
process by persisting the `DECOMMISSIONED` status.

When transitioning between `MembershipStatus`es, we CPut against
what's already there, disallowing illegal state transitions. The
appropriate error codes are surfaced back to the user. An example would
be in attempting to recommission a fully decommissioned node, in which
case we'd error out with the following:

> ERROR: can only recommission a decommissioning node; n4 found to be
> decommissioned

Note that this is a behavioral change for `cockroach node recommission`.
Previously it was able to recommission any "fully decommissioned" node,
regardless of how long ago it's was removed from the cluster. Now
recommission serves to only cancel an accidental decommissioning process
that wasn't finalized.

The `decommissioning` column in `crdb_internal.gossip_liveness` is now
powered by this new `MembershipStatus` instead, and we introduce a new
`membership` column to it that should be preferred going forward. We
also introduce the same column to the output generated by `cockroach
node status --decommission`. The `is_decommissioning` column still exists,
but is also powered by this `MembershipStatus`.

While here, we iron out the events plumbed into `system.eventlog`: it
now has a dedicated event for "node decommissioning".

---

Release note (general change): `cockroach node recommission` has new
semantics. Previously it was able to recommission any decommissioning node,
regardless of how long ago it's was decommissioned, or removed from the
cluster. Now recommission serves to only cancel an accidental
inflight decommissioning process that wasn't finalized.

Release note (cli change): We introduce a `membership` column to the
output generated by `cockroach node status --decommission`. It should be
used in favor of the `is_decommissioning` column going forward.

Release note (cli change): The v20.2 cli `cockroach node` family of
subcommands will not work with against servers running older version of
cockroach, but the v20.1 cli `cockroach node` subcommands will work
against v20.2 servers.

Release note (cli change): The `is_decommissioning` column found in the
output of `cockroach node decommission` is slated for removal in v20.1.
Operators should instead use the new `membership` column to determine
node membership status.

irfansharif committed:
roachtest: improve decommissioning roachtests

Re-write the previously skipped `acceptance/decommission` to
account for new node {d,r}ecommissioning semantics. The minimum version
to run this test against is v20.2.

```go
// runDecommissionRecommission tests a bunch of node
// decommissioning/recommissioning procedures, all the while checking for
// replica movement and appropriate membership status detection behavior. We go
// through partial decommissioning of random nodes, ensuring we're able to undo
// those operations. We then fully decommission nodes, verifying it's an
// irreversible operation.
```

Release note: None

irfansharif committed:
roachtest: add `decommission/mixed-versions`

Add a roachtest stressing randomized `cockroach node
{decommission,recommission}` usage in multi-version clusters.

Release note: None

irfansharif committed:
kvserver: rename members of `livenessUpdate`

Pulling out a rename refactor out of future commits that then make use of it.

Release note: None

irfansharif reviewed (commented):
Thanks for the review, Tobi. I ended up re-writing most of what so I'd just look the last few commits. (It's basically a new PR, and partly why I've been dragging my feet with this one.) 

Also added randomized multi-version and regular decomm/recomm tests.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-:-MBXOAwtEkT5CskJYAkb:b-lp3nd9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @tbg)

___
*[pkg/cli/demo_cluster.go, line 375 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5nA915hpzgTPRijuU:-MAYSZW14MwK9TJspngq:bhmdwb6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/demo_cluster.go#L375)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

why `Form`? Isn't `CommissionStatusFromBool` good enough?
</blockquote></details>

Done (ended up removing entirely).
___
*[pkg/cli/node.go, line 66 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5nG9aAl2XQtT5_C0L:-MAYTLS0--zdymVlwAvK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L66)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Add a release note that the 20.2 cli `node` family of commands will not work with older versions, but that older versions will work against 20.2.
</blockquote></details>

Done.
___
*[pkg/cli/node.go, line 340 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5oCyfBjNiAR6UDeJE:-MAYTVP56UaKdJB7s0yh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L340)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

If a 20.2 cli were used against a 20.1 node, it would always recommission, which is bad.
I think this problem goes away if you re-use the field as I suggested in another comment.
</blockquote></details>

Done.
___
*[pkg/cli/node.go, line 361 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5oxkX88YCveznVAeI:-MAYVFhWDuDk2e5VLU80:b-2l44la) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L361)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Uh, oh. `false && false || true == true`, so this is a bug. You want `false && (false || true)`
</blockquote></details>

Done (re-wrote to be more readable).
___
*[pkg/cli/node.go, line 364 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5pOCQ4t5D7hGdzp7x:-MAYVGCL2S9i5artqc1_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L364)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

You would never want to enter this if `!allDecommissioning`, no?
</blockquote></details>

Done.
___
*[pkg/cli/node.go, line 365 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5ojCa4GgbWgjM1snT:-MAYVGwyEvMjdjgET8Lh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L365)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

node(s)
</blockquote></details>

Done.
___
*[pkg/cli/node.go, line 389 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5pGvODpM2GgrIgy5n:-MB8Gl6e0hf745ib-fkD:b-h9gmix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L389)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

But you may already have done it? I think you need to make sure you never do it in that case.

Let's give this method a real close look, since it seems really easy to introduce bugs here.
</blockquote></details>

The comment below was off. I've reworked this method a bit to make it clear what the intent was, PTAL.
___
*[pkg/cli/node.go, line 419 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5podeEyakZx_SE4hw:-MAYWHv6BrZoA11n1hz6:b5ypr5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L419)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

All of these too will be better off with the enum reusing the bool field. Right now if you use a new cli against old node you'll just get wrong answers. I'm ok breaking compatibility here but we shouldn't silently return wrong data. Problem will disappear with enum reuse, luckily.
</blockquote></details>

Done (yikes, didn't even notice this one).
___
*[pkg/cli/node.go, line 446 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5qK380yR6ENzPK5Kl:-MAZG9gx3xqWM_pvCNPh:bnhm957) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L446)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This returns -1 if not found, in which case you'll explode below.

I see now that you guard against that at the caller, but move that check in here. If you can't match, return "".
</blockquote></details>

Removed entirely.
___
*[pkg/cli/node.go, line 474 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5qd3v34cfDlOpUXsj:-MAZFhoiBR7AtbRonnkl:bdvh0r1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/node.go#L474)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Isn't there a better way to solve this? You can return a structured grpc error:

```go
err := status.Error(codes.FailedPrecondition, "something something blah blah")
```

and then later

```
if s, ok := status.FromError(err); ok && s.Code() == codes.FailedPrecondition {
	fmt.Println(s.Message())
}
```
</blockquote></details>

Done (TIL, my first attempt here was to look at the cockroachdb/errors package but it seems that we never unflatten errors off the wire, so didn't know what else to try).
___
*[pkg/cli/interactive_tests/test_demo_node_cmds.tcl, line 70 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5sCUf7JqtHv8gMpON:-MAZGLB-ASTf9TUA1PcZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cli/interactive_tests/test_demo_node_cmds.tcl#L70)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It's been lingering in my mind throughout the review, and I think for a good reason - I'm not sure `commissioned` really is the most intuitive choice here. It looks like a verb - like we commissioned a node, i.e. we ordered one, it's bound to arrive any time soon. I feel that this might be confusing to users especially when we have some very obvious choices available, like `active`. Consider switching to that.
</blockquote></details>

Done.
___
*[pkg/cli/interactive_tests/test_multiple_nodes.tcl, line 24 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5mS6H0d-BqHDeG8cN:-MAZGQUmEqBYfOGqERov:bjbgvp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/af0f7bb31be91b0fe61a30d82054223293c6c597/pkg/cli/interactive_tests/test_multiple_nodes.tcl#L24)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Do we want to maintain these warnings though? In that case it seems wise to remove the code only when we're ready to replace it with new code. It seems that this PR is trying to do two things at once - add the decommission bit and deprecating the liveness status thing, but it's not really following through on the latter, so I would prefer if we kept that to the minimum and followed through in a separate PR.
</blockquote></details>

Separated it out in https://github.com/cockroachdb/cockroach/pull/50478.
___
*[pkg/cmd/roachtest/acceptance.go, line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5sh4r1Bx1hAigt-Bx:-MAZGUnu6xZcO0Ely6s1:bj3arkd) ([raw file](https://github.com/cockroachdb/cockroach/blob/af0f7bb31be91b0fe61a30d82054223293c6c597/pkg/cmd/roachtest/acceptance.go#L38)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

?
</blockquote></details>

The test was moved into decommission.go, and renamed.
___
*[pkg/cmd/roachtest/decommission.go, line 253 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5t_wpDRVEUBJq6tqV:-MAZH-Z11N5MJDV61qcH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cmd/roachtest/decommission.go#L253)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

There's no issue open for it, so let's rename it now.
</blockquote></details>

Done.
___
*[pkg/cmd/roachtest/decommission.go, line 279 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5tk0Z10G-FgJ0_lnt:-MB8H8Yz-zRYIcURhQ3x:b-5lu4m5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/cmd/roachtest/decommission.go#L279)):*
Reworked this test to be a bit clearer, and added multi-node decommissioning checks (+ comments). 

> Did you verify that this still works against <= 20.1?

I don't follow, why would this need to? Wouldn't we run the version of this roachtest present in 20.1 code to run against 20.1 code?
___
*[pkg/kv/kvserver/node_liveness.go, line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5uVFC5MYbOemAlR9Q:-MB8Hfdg9zJ7jEi5gfDF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/kv/kvserver/node_liveness.go#L49)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Similar concern about `commissioning status` (which pops up in a number of column headers). What about `membership` status? After all, decommissioning is all about removing members from a cluster and in the context of long-running operations, we are trying to reason about cluster membership in an even stricter sense (who *can* access the cluster as opposed to who is tracked as a member)
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/node_liveness.go, line 360 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5vN5wDZX07su7gHiZ:-MB8HyvR6CUWzfzHY84y:bqf2diq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/kv/kvserver/node_liveness.go#L360)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This shouldn't fail in mixed-version clusters. I don't think it would when you take my suggestion to merge the fields. The only caveat is that we are not allowed to mark as `decommissioned` but the `cli` should just special-case the mixed version error there and print a warning to the client (but succeed anyway).
</blockquote></details>

Merged the fields, and succeed anyway now in mixed version cluster settings.
___
*[pkg/kv/kvserver/node_liveness.go, line 382 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5vZlJ7z0frDYGF8rF:-MAZIDD7CBZanfYUlEeI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/kv/kvserver/node_liveness.go#L382)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Let's
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/node_liveness.go, line 395 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5viwi6YJg2NEQ9H3x:-MAZIFEO1WueKhEtT7Eg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/kv/kvserver/node_liveness.go#L395)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Just a heads up that I'm going to skip lightly over this mixed version stuff for now since I expect it to change.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/node_liveness.go, line 438 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5vpBGD1l96sAzynBd:-MAZItp18ciKeusIRca1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/kv/kvserver/node_liveness.go#L438)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

`new` is also a keyword, let's avoid that.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/kvserverpb/liveness.go, line 71 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5wEBODVbUDQ9EgcRK:-MAZIydPCoUlUhBcdKDd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/kv/kvserver/kvserverpb/liveness.go#L71)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Could this really be removed? Nodes that never join the cluster with 20.2 will never update their liveness, so you'd have to retain some understanding of the old format ~forever (or write an explicit migration).
I think this problem will go away when you merge the fields.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/kvserverpb/liveness.proto, line 51 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5kZYK0J_q096PRfsd:-MAZJ6_S3SMrSLvO9_6k:b2x8jmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cb13324e4f721178bbe942958e3c3d5dfa05b96/pkg/kv/kvserver/kvserverpb/liveness.proto#L51)):*
Done. This cleaned a lot of things up, I didn't realize it was possible. I was looking at https://developers.google.com/protocol-buffers/docs/proto#updating:

> enum is compatible with int32, uint32, int64, and uint64 in terms of wire format 

And took it at face value given that bool was not explicitly mentioned. But I see right above:

> int32, uint32, int64, uint64, and bool are all compatible

Would've saved myself a lot of headache.
___
*[pkg/kv/kvserver/kvserverpb/liveness.proto, line 128 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5l8x2B5PAWGcoYEVT:-MAZJfLL10JPKwl9vSu_:bh4kmxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cb13324e4f721178bbe942958e3c3d5dfa05b96/pkg/kv/kvserver/kvserverpb/liveness.proto#L128)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Next time, make sure you put pure renames like that in a separate commit. There were three things going in on in this one which was mildly unfortunate.
</blockquote></details>

Did you mean separate PR? I had pulled in the proto renames in the first commit, but agreed that the NodeLivenessStatus deprecation was entirely orthogonal (https://github.com/cockroachdb/cockroach/pull/50478).
___
*[pkg/server/server.go, line 1807 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5weOUBySENylSRirM:-MAZKukF8MWP4-yyRlAN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/server/server.go#L1807)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

And here you'd check the cluster version.
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 1809 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5wgc94GArAcHcx-k-:-MAZL64W2xfb4QF2I8eI:bkryyu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/server/server.go#L1809)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

We need to avoid panicking whenever we can. Here it's easy to return an error.
</blockquote></details>

This seems like a panic-worthy assertion, no? There are only three possible states defined for the targetStatus, and if we were to introduce more, we should like to think about the events that would correspond to (so, programming error if this was missed).
___
*[pkg/server/serverpb/admin.proto, line 879 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5lTKcDdW1SN1jRbgV:-MAZOtH87zmg5qcSgjtE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cb13324e4f721178bbe942958e3c3d5dfa05b96/pkg/server/serverpb/admin.proto#L879)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I would prefer if we just stuck with the old name throughout. `CommissionRequest` sounds like a request to "commission" a node, i.e. to mark it as commissioned, and thus doesn't really seem preferable to `DecommissionRequest` (which at least will be used for its name most of the time).
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 2742 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5wpnL7a23YKOocdOb:-MAZP1kK2-7iYxGesXG1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/sql/crdb_internal.go#L2742)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Looking at this, I think `membership: active` would have a good intuitive ring to it.
</blockquote></details>

Done.
___
*[pkg/sql/event_log.go, line 90 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50329#-MA5wzac8bircHsWKSzQ:-MB8IGsOArmbdMJAl_VQ:bkrqwwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f222ac2584b2d33fd42014219dbbe4ffadbdff6c/pkg/sql/event_log.go#L90)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Is there a migration concern with this change? Not really right? They're only ever used as strings.
</blockquote></details>

Not as far as I can tell.


<!-- Sent from Reviewable.io -->

irfansharif commented:
@knz, this is ready for you to take a look as well.

irfansharif commented:
I just learned that roachtests need to be compatible across multiple versions of crdb since they're built from master in our CI, and run against earlier version builds. The original `acceptance/decommission` test was backwards compatible, but my newer randomized `decommission-recommission` isn't. I've thrown in a
```
MinVersion: "v20.2.0",
```
in there for now given the test was written with the new recommissioning behavior in mind, but now we've lost decommission test coverage against v20.1/v19.2/v19.1 builds. Is that fine? I can revive the earlier version of `acceptance/decommission` and define a `MaxVersion` in the test spec to ignore v20.2 builds, if asked (probably in a separate PR).

tbg commented:
> but now we've lost decommission test coverage against v20.1/v19.2/v19.1 builds

The acceptance/decommission tests aren't our only coverage, we also have `decommission/...`.
Given that and given how difficult failures of acceptance/decommission are to debug (I will try to suggest some improvements now that you're touching the code) I am ok with sacrificing this test on the release branches.

tbg commented on pkg/cmd/roachtest/decommission.go:
```diff
@@ -257,34 +273,92 @@ func execCLI(
 }
 
-func runDecommissionAcceptance(ctx context.Context, t *test, c *cluster) {
+// runDecommissionRecommission tests a bunch of node
```
(can't comment on line 271 above, but that's what I'm talking about)

Looking into failures of this test recently, I think I saw that the output of failing decommissioning commands was nowhere to be found. I think this was because of `RunWithBuffer` - output doesn't go to artifacts, so the only place we get it is `buf`, but of course we don't typically print `buf` if we get an `err`.

Could you try to use something like this:

```go
	node := c.Node(runNode)
	l, _ := c.l.ChildLogger(cmdLogFileName(timeutil.Now(), node, args...)
	err := c.RunL(ctx, l, node, args...)
	l.close()
	out, _ := ioutil.ReadAll(l.path)
	return string(out), err
```

The difference is that we get the output in a file 100% of the time, so if a decommission invocation just hangs, we'll see where.

tbg commented on pkg/cmd/roachtest/decommission.go:
```diff
@@ -257,34 +273,92 @@ func execCLI(
+
+	// We define a few short hands to {d,r}ecommission target nodes, through a
+	// given node.
 	decommission := func(
```
Could you give this test the "stateful doover"? Add a `type decomTestHelper struct { ... }` and put everything in it that is now referenced by all of these closures, which then become anchored to `*decomTestHelper`.

This allows you to move all of the boilerplate out of the main test, and in particular *below* it (so that readers see the test first and don't have to wade through the setup unless they want to).

tbg commented on pkg/cmd/roachtest/decommission.go:
```diff
@@ -326,41 +400,44 @@ func runDecommissionAcceptance(ctx context.Context, t *test, c *cluster) {
+		"gossiped_replicas", "is_decommissioning", "membership", "is_draining",
 	}
-	getStatusCsvOutput := func(ids []string, numCols int) [][]string {
+	// Index of `membership` column in statusHeaderWithDecommission
```
.

tbg commented on pkg/cmd/roachtest/decommission.go:
```diff
@@ -257,34 +273,92 @@ func execCLI(
+		return len(records)
 	}
 
+	// matchCsv matches a multi-line csv string with the provided regex
```
That we do all of this matching here is sort of crazy and makes this test a real mess. Ideally we'd have the decommission command a super thin wrapper around hitting a gRPC endpoint. Definitely not something you can just drive-by here though.

Curious if @knz has an intermediate idea. The hoops we jump through in this test are crazy, no? This test would be so much less messy if we just cared about the exit status of the invocation rather than its output.

tbg commented on pkg/cmd/roachtest/decommission.go:
```diff
@@ -374,318 +451,486 @@ func runDecommissionAcceptance(ctx context.Context, t *test, c *cluster) {
-		t.Fatal(err)
-	}
+		// Check that `node status` reflects an ongoing decommissioning status
+		// for the second node.
```
This won't flake, right? The decommission status is propagates async, but behind the scenes we always hit KV here?

irfansharif replied:
Yup.

tbg commented on pkg/cmd/roachtest/decommission.go:
```diff
@@ -374,318 +451,486 @@ func runDecommissionAcceptance(ctx context.Context, t *test, c *cluster) {
-		if err != nil {
+		runNode := getRandNode()
+		t.l.Printf("attempting to decommission all nodes from n%d\n", runNode)
+		if err := retry.WithMaxAttempts(ctx, retryOpts, 50, func() error {
```
Why would this fail?

tbg commented on pkg/cmd/roachtest/decommission.go:
```diff
@@ -374,318 +451,486 @@ func runDecommissionAcceptance(ctx context.Context, t *test, c *cluster) {
-		t.l.Printf("re-recommissioning first node, from itself\n")
-		if _, err := decommission(ctx, 1, nil, "recommission", "--self"); err != nil {
-			t.Fatalf("recommission failed: %v", err)
+		// Cancel in-flight decommissioning process of target node.
```
all nodes

tbg commented on pkg/cmd/roachtest/decommission.go:
```diff
@@ -374,318 +451,486 @@ func runDecommissionAcceptance(ctx context.Context, t *test, c *cluster) {
-			"decommission", "--format=csv") // wait=all is implied
-		if err != nil {
-			t.Fatalf("decommission failed: %v", err)
+		// Check that even though the node is decommissioned, we still see it (since
```
even though two nodes are decommissioned, we still see them (since they remain live)

tbg commented on pkg/cmd/roachtest/decommission.go:
```diff
@@ -374,318 +451,486 @@ func runDecommissionAcceptance(ctx context.Context, t *test, c *cluster) {
+			t.l.Printf("expected to fail: recommissioning [n%d,n%d] (from n%d)\n",
+				targetNodeA, targetNodeB, runNode)
+			if _, err := recommission(ctx, c.Nodes(targetNodeA, targetNodeB), runNode); err == nil {
+				t.Fatal("expected recommission to failed")
```
fail

tbg commented on pkg/cmd/roachtest/decommission.go:
```diff
@@ -374,318 +451,486 @@ func runDecommissionAcceptance(ctx context.Context, t *test, c *cluster) {
+		{
+			runNode = getRandNode()
+			t.l.Printf("checking that decommissioning [n%d,n%d] (from n%d) is a no-op\n",
+				runNode, targetNodeA, targetNodeB)
```
runNode needs to go last

tbg commented on docs/generated/settings/settings.html:
```diff
@@ -72,6 +72,6 @@
 <tr><td><code>trace.lightstep.token</code></td><td>string</td><td><code></code></td><td>if set, traces go to Lightstep using this token</td></tr>
 <tr><td><code>trace.zipkin.collector</code></td><td>string</td><td><code></code></td><td>if set, traces go to the given Zipkin instance (example: '127.0.0.1:9411'); ignored if trace.lightstep.token is set</td></tr>
-<tr><td><code>version</code></td><td>custom validation</td><td><code>20.1-9</code></td><td>set the active cluster version in the format '<major>.<minor>'</td></tr>
+<tr><td><code>version</code></td><td>custom validation</td><td><code>20.1-10</code></td><td>set the active cluster version in the format '<major>.<minor>'</td></tr>
```
Just so that I don't forget, we are making sure to never persist a state that's not zero (false) or one (true) until that version is active, right?

tbg commented on pkg/cli/node.go:
```diff
@@ -64,9 +66,10 @@ func runLsNodes(cmd *cobra.Command, args []string) error {
 		conn,
 		makeQuery(`SELECT node_id FROM crdb_internal.gossip_liveness
-               WHERE decommissioning = false OR split_part(expiration,',',1)::decimal > now()::decimal`),
+               WHERE membership = 'active' OR split_part(expiration,',',1)::decimal > now()::decimal`),
```
Is the old column still around? I.e. will `./cockroach-20.1 node ls` work against 20.2 cluster? I think the answer is yes, just checking.

tbg commented on pkg/cli/error.go:
```diff
@@ -268,6 +268,14 @@ func MaybeDecorateGRPCError(
+				// The v20.2 binary makes use of columns not present in v20.1,
+				// so this is a disallowed operation. Surface a better error
+				// code here.
+				return fmt.Errorf("cannot use a v20.2 cli against servers running v20.1")
```
cc @knz just to make sure this is ok - we want the 20.2 binary to use this new column name (the old one stays around for 20.1 usage). Or is there a simple version check we can use to determine the column to use? EIther way, if you feel like we don't need to support it, I'd prefer to keep this simple.

tbg commented on pkg/kv/kvserver/kvserverpb/liveness.go:
```diff
@@ -41,10 +43,66 @@ func (l *Liveness) IsDead(now time.Time, threshold time.Duration) bool {
+
+func (c MembershipStatus) String() string {
+	switch c {
+	case MembershipStatus_ACTIVE:
```
Add a comment that these must not be changed since the cli matches on them.

tbg commented on pkg/kv/kvserver/kvserverpb/liveness.go:
```diff
@@ -41,10 +43,66 @@ func (l *Liveness) IsDead(now time.Time, threshold time.Duration) bool {
+			new.NodeID, old.Membership.String())
+		return status.Error(codes.FailedPrecondition, err)
+	}
+	if new.Membership.Decommissioning() && !old.Membership.Active() {
```
But decommissioning is still idempotent? Unclear to me when this is called then

irfansharif replied:
This is actually dead code, I had comment to that effect earlier but removed it. We no-op away operations at the caller when observed state is the same as the target state.

```go
	} else if oldLivenessRec.Membership.Decommissioned() &&
		newLiveness.Membership.Decommissioning() {
		// Marking a decommissioned node for decommissioning is a no-op.
		return false, nil
	}
```

Re-wrote the function a bit to encode this specific no-op transition.

tbg commented on pkg/kv/kvserver/node_liveness.go:
```diff
@@ -424,35 +436,69 @@ type livenessUpdate struct {
+	oldLivenessRec LivenessRecord,
+	targetStatus kvserverpb.MembershipStatus,
+) (statusChanged bool, err error) {
+	// Lets compute what our new liveness record should be.
```
's

tbg commented on pkg/kv/kvserver/node_liveness.go:
```diff
@@ -424,35 +436,69 @@ type livenessUpdate struct {
+	if oldLivenessRec.Liveness == (kvserverpb.Liveness{}) {
+		// Liveness record didn't previously exist, so we create one.
+		//
+		// ASK(reviewers): Why is it possible for a liveness record to not exist
```
It is just generally difficult to write it at startup. When a node joins the cluster, this completes before it has had a chance to write its liveness record. If it gets decommissioned immediately, there won't be one yet. The Connect RPC can solve this though, I think? We can bootstrap clusters with a liveness record for n1. Any other node at some point has to join the cluster for the first time via the Connect RPC, which as part of its job can make sure the liveness record exists before responding to the new node.

tbg commented on pkg/kv/kvserver/node_liveness.go:
```diff
@@ -730,7 +783,7 @@ func (nl *NodeLiveness) GetIsLiveMap() IsLiveMap {
 	for nID, l := range nl.mu.nodes {
 		isLive := l.IsLive(now)
-		if !isLive && l.Decommissioning {
+		if !isLive && l.Membership.Decommissioning() || l.Membership.Decommissioned() {
```
Add parentheses please

tbg commented on pkg/kv/kvserver/node_liveness.go:
```diff
@@ -1009,12 +1064,14 @@ func (nl *NodeLiveness) maybeUpdate(new LivenessRecord) {
 	}
 }
 
+// shouldReplaceLiveness checks to see if the new liveness, is infact, newer
```
if the new liveness is in fact newer

tbg commented on pkg/kv/kvserver/replica_gossip.go:
```diff
@@ -176,6 +177,15 @@ func (r *Replica) MaybeGossipNodeLiveness(ctx context.Context, span roachpb.Span
+			// representation unless we're told by the user that there are no
+			// pre-v20.1 nodes around. We should never get here.
+			if kvLiveness.Membership.Decommissioned() {
+				panic("programming error: illegal membership status: decommissioned")
```
fatal

tbg commented on pkg/cmd/roachtest/mixed_version_decommission.go:
```diff
@@ -0,0 +1,287 @@
+
+// runDecommissionMixedVersions runs through randomized
+// decommission/recommission processes in mixed-version clusters.
+func runDecommissionMixedVersions(
```
Move this to the top, as it's the best entry point for reading what this test is about.

tbg commented on pkg/cmd/roachtest/mixed_version_decommission.go:
```diff
@@ -0,0 +1,287 @@
+
+	// The v20.2 CLI can only be run against servers running v20.2. For this
+	// reason, we grab a handle on a specific server slated for an upgrade.
+	pinnedUpgrade := getRandNode()
```
Is this randomness useful? If this test flakes, it will require more cognitive overhead to track failure modes and to poke at reproductions.

tbg commented on pkg/cmd/roachtest/acceptance.go:
```diff
@@ -35,13 +35,6 @@ func registerAcceptance(r *testRegistry) {
 		{name: "build-analyze", fn: runBuildAnalyze},
 		{name: "cli/node-status", fn: runCLINodeStatus},
-		{
-			name: "decommission", fn: runDecommissionAcceptance,
```
but isn't the code still there?

tbg reviewed (approved):
Phew, that was a chunk of work, I bet you'll be glad when it's in.

When the review settles down and you get ready for merging, could you run a good number of iterations (100?) of the roachtests to shake out flakes? And if you see any flakes, make sure they are not opaque (i.e. output nowhere to be found, etc).

irfansharif reviewed (commented):
> could you run a good number of iterations (100?) of the roachtests to shake out flakes

Tried it out this morning, nothing fell out. Merging on green, TFTR!

irfansharif commented:
Flaked on https://github.com/cockroachdb/cockroach/issues/51179 I think.

```
panic: flow: unexpected 10240 leftover bytes [recovered]
	panic: flow: unexpected 10240 leftover bytes
```

Merge conflict with #51168. Retrying..

irfansharif commented:
bors r=tbg
